### PR TITLE
[url_launcher]added windowName parameter in url_launcher

### PR DESCRIPTION
**Pull request by balvinderz**
## Description

Adds the ability to launch a url in same tab,new tab or launch the  url in an  iframe.

## Related Issues

https://github.com/flutter/flutter/issues/56867

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [X] I read and followed the [Flutter Style Guide].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I signed the [CLA].
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [X] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by balvinderz** [comment]
@mklim @ditman

File: packages/url_launcher/url_launcher/lib/url_launcher.dart [hunk 1]
```diff
@@ -93,6 +97,7 @@ Future<bool> launch(
     enableDomStorage: enableDomStorage ?? false,
     universalLinksOnly: universalLinksOnly ?? false,
     headers: headers ?? <String, String>{},
+    windowName: windowName ?? "_blank",
```
Comment (ditman) [inline packages/url_launcher/url_launcher/lib/url_launcher.dart:100]: `"_blank"` only makes sense in the web. Please, do not set `"_blank"` as a default value here, otherwise this will break the logic to auto-determine the target on the web version.

File: packages/url_launcher/url_launcher/pubspec.yaml [hunk 2]
```diff
@@ -22,13 +22,21 @@ flutter:
 dependencies:
   flutter:
     sdk: flutter
-  url_launcher_platform_interface: ^1.0.4
+  url_launcher_platform_interface:
+    git:
+      url: https://github.com/balvinderz/plugins
+      ref: windowname
+      path: packages/url_launcher/url_launcher_platform_interface
```
Comment (ditman) [inline packages/url_launcher/url_launcher/pubspec.yaml:29]: This makes the plugin unpublishable. Once we figure out all the rest of the changes, you'll have to split this PR in 3: one for each of the packages you've touched (and merge them in the correct order)

File: packages/url_launcher/url_launcher/test/url_launcher_test.dart [hunk 3]
```diff
@@ -44,7 +44,7 @@ void main() {
     });
 
     test('default behavior', () async {
-      await launch('http://flutter.dev/');
+      await launch('http://flutter.dev/', windowName: "_blank");
```
Comment (ditman) [inline packages/url_launcher/url_launcher/test/url_launcher_test.dart:47]: Passing windowName is *not* the default behavior, please undo this test. Ideally you should add another test of what happens when you pass `windowName`.

File: packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart [hunk 4]
```diff
@@ -35,6 +35,7 @@ class MethodChannelUrlLauncher extends UrlLauncherPlatform {
     @required bool enableDomStorage,
     @required bool universalLinksOnly,
     @required Map<String, String> headers,
+    @required String windowName,
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart:38]: Since windowName is web-only, it can't be `@required`.

File: packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart [hunk 5]
```diff
@@ -47,15 +47,14 @@ abstract class UrlLauncherPlatform extends PlatformInterface {
...
-    @required Map<String, String> headers,
-  }) {
+  Future<bool> launch(String url,
+      {@required bool useSafariVC,
+      @required bool useWebView,
+      @required bool enableJavaScript,
+      @required bool enableDomStorage,
+      @required bool universalLinksOnly,
+      @required Map<String, String> headers,
+      @required String windowName}) {
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart:57]: `windowName` must be optional, not `@required`. Also, to follow Flutter convention, this'd be better called: `webOnlyLinkTarget` (or similar, so people know at a first glance that they'll set the target attribute of a link in web only).

File: packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart [hunk 6]
```diff
@@ -46,6 +47,7 @@ class MethodChannelUrlLauncher extends UrlLauncherPlatform {
         'enableDomStorage': enableDomStorage,
         'universalLinksOnly': universalLinksOnly,
         'headers': headers,
+        'windowName': windowName,
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart:50]: Since this is a web-only parameter, there's no need for passing it through the Method Channel (which is only used by the Android/iOS implementations). Do not add 'windowName' to the Map that is passed to invokeMethod.

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 7]
```diff
@@ -39,15 +39,15 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
...
   ///
   /// Returns the newly created window.
   @visibleForTesting
-  html.WindowBase openNewWindow(String url) {
+  html.WindowBase openNewWindow(String url, {windowName = ""}) {
     // We need to open mailto, tel and sms urls on the _top window context on safari browsers.
     // See https://github.com/flutter/flutter/issues/51461 for reference.
     final target =
-        browser.isSafari && _isSafariTargetTopScheme(url) ? '_top' : '';
+        browser.isSafari && _isSafariTargetTopScheme(url) ? '_top' : windowName;
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:50]: `windowName` can be null, so please ensure it's `windowName ?? ''`

File: packages/url_launcher/url_launcher_web/test/url_launcher_web_test.dart [hunk 8]
```diff
@@ -110,6 +113,7 @@ void main() {
               enableDomStorage: null,
               enableJavaScript: null,
               headers: null,
+              windowName: "_blank",
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/url_launcher_web_test.dart:116]: Revert this change, windowName should not be "_blank" by default. Also, add new tests to the `openNewWindow` group below, to test what happens when different values of windowName are passed in.

**Comment by balvinderz** [comment]
@ditman  i did the suggested changes.

File: packages/url_launcher/url_launcher/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,6 @@
+## 5.5.1
+* Added windowName web only parameter
```
Comment (ditman) [inline packages/url_launcher/url_launcher/CHANGELOG.md:2]: Please leave a space between the version and the list. Also, let's call the parameter:

`webOnlyWindowName`

(The parameter is used as the "windowName" of the `window.open` function, so I think that's the best name for now!)

File: packages/url_launcher/url_launcher/lib/url_launcher.dart [hunk 10]
```diff
@@ -63,6 +66,7 @@ Future<bool> launch(
   bool universalLinksOnly,
   Map<String, String> headers,
   Brightness statusBarBrightness,
+  String webOnlyLinkTarget,
```
Comment (ditman) [inline packages/url_launcher/url_launcher/lib/url_launcher.dart:69]: Let's call the parameter:

`webOnlyWindowName` everywhere.

File: packages/url_launcher/url_launcher_platform_interface/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,7 @@
+## 1.0.8
+
+* Added windowName web only parameter
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/CHANGELOG.md:3]: Let's call the parameter:

`webOnlyWindowName` everywhere.

File: packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart [hunk 12]
```diff
@@ -27,15 +27,14 @@ class MethodChannelUrlLauncher extends UrlLauncherPlatform {
...
-    @required Map<String, String> headers,
-  }) {
+  Future<bool> launch(String url,
+      {@required bool useSafariVC,
+      @required bool useWebView,
+      @required bool enableJavaScript,
+      @required bool enableDomStorage,
+      @required bool universalLinksOnly,
+      @required Map<String, String> headers,
+      String webOnlyLinkTarget}) {
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/lib/method_channel_url_launcher.dart:37]: Let's call the parameter:

`webOnlyWindowName` everywhere.

Also: add a comma after the last named parameter, so the `dartfmt` looks more like the previous version :)

File: packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart [hunk 13]
```diff
@@ -47,15 +47,14 @@ abstract class UrlLauncherPlatform extends PlatformInterface {
...
-    @required Map<String, String> headers,
-  }) {
+  Future<bool> launch(String url,
+      {@required bool useSafariVC,
+      @required bool useWebView,
+      @required bool enableJavaScript,
+      @required bool enableDomStorage,
+      @required bool universalLinksOnly,
+      @required Map<String, String> headers,
+      String webOnlyLinkTarget}) {
```
Comment (ditman) [inline packages/url_launcher/url_launcher_platform_interface/lib/url_launcher_platform_interface.dart:57]: Let's call the parameter:

`webOnlyWindowName` everywhere.

Also: add a comma after the last named parameter, so the `dartfmt` looks more like the previous version :)

File: packages/url_launcher/url_launcher_web/CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,7 @@
+# 0.1.3
+
+- Added "windowName" parameter that lets you launch url in new tab,same tab or in an iframe
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/CHANGELOG.md:3]: Let's call the parameter:

`webOnlyWindowName` everywhere.

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 15]
```diff
@@ -34,20 +34,25 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
     UrlLauncherPlatform.instance = UrlLauncherPlugin();
   }
 
-  String _getUrlScheme(String url) => Uri.tryParse(url)?.scheme;
+  String _getUrlScheme(String url) =>
+      Uri
+          .tryParse(url)
+          ?.scheme;
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:40]: Can you move this all back to a single line? I don't understand why this got reformatted like that

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 16]
```diff
@@ -34,20 +34,25 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
...
 
   bool _isSafariTargetTopScheme(String url) =>
       _safariTargetTopSchemes.contains(_getUrlScheme(url));
 
-  /// Opens the given [url] in a new window.
+  /// Opens the given [url] in the specified [webOnlyLinkTarget].
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:45]: Let's call the parameter:

`webOnlyWindowName` everywhere.

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 17]
```diff
@@ -34,20 +34,25 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
...
   ///
   /// Returns the newly created window.
   @visibleForTesting
-  html.WindowBase openNewWindow(String url) {
+  html.WindowBase openNewWindow(String url,{webOnlyLinkTarget} ) {
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:49]: Let's call the parameter:

`webOnlyWindowName` everywhere.

Also: this is not correctly `dartfmt`, so the formatting tests are failing.

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 18]
```diff
@@ -34,20 +34,25 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
...
     // We need to open mailto, tel and sms urls on the _top window context on safari browsers.
     // See https://github.com/flutter/flutter/issues/51461 for reference.
     final target =
-        browser.isSafari && _isSafariTargetTopScheme(url) ? '_top' : '';
+    browser.isSafari && _isSafariTargetTopScheme(url)
+        ? '_top'
+        : webOnlyLinkTarget ?? "";
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:55]: Reading the [original ticket](https://github.com/flutter/flutter/issues/56867#issuecomment-629363114), I think this logic is backwards:

* If the user is passing a webOnlyLinkTarget (!= null), use that
* Else, use the older logic

So I guess this needs to be:

```dart
final target = webOnlyLinkTarget ?? browser.isSafari && _isSafariTargetTopScheme(url) ? '_top' : '';
```

File: packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart [hunk 19]
```diff
@@ -57,15 +62,16 @@ class UrlLauncherPlugin extends UrlLauncherPlatform {
...
     @required bool useWebView,
     @required bool enableJavaScript,
     @required bool enableDomStorage,
     @required bool universalLinksOnly,
     @required Map<String, String> headers,
+    String webOnlyLinkTarget,
   }) {
-    return Future<bool>.value(openNewWindow(url) != null);
+    return Future<bool>.value(
+        openNewWindow(url,webOnlyLinkTarget: webOnlyLinkTarget) != null);
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart:75]: dartfmt is probably complaining about this too

File: packages/url_launcher/url_launcher_web/test/url_launcher_web_test.dart [hunk 20]
```diff
@@ -146,6 +145,20 @@ void main() {
...
       });
 
+      test('setting webOnlyLinkTarget as _self opens the url in the same tab',
+          () {
+        plugin.openNewWindow("https://www.google.com",
+            webOnlyLinkTarget: "_self");
+        verify(mockWindow.open('https://www.google.com', '_self'));
+      });
+
+      test('setting webOnlyLinkTarget as _blank opens the url in a new tab',
```
Comment (ditman) [inline packages/url_launcher/url_launcher_web/test/url_launcher_web_test.dart:155]: This is exactly what needs to be tested, thanks!

PS: Once you change the logic above (use whatever the user passed instead of the automatic detection), can you add a new test in the `Safari` group below to see that if the user passes a '_blank' to a `mailto:` url, the plugin will use '_blank' instead of '_top'?

**Comment by balvinderz** [comment]
Done with the changes

File: packages/url_launcher/url_launcher_web/CHANGELOG.md (hunk 14 above)
Comment (ditman) [reply packages/url_launcher/url_launcher_web/CHANGELOG.md:3]: Please update this doc! :)

**Comment by ditman** [comment]
I think this looks good now, but our tooling doesn't allow us to merge multiple plugins at once when they depend amongst themselves.

Can you prepare a PR with only the changes to the url_launcher_platform_interface package? Once we publish that one, we should be able to publish the rest!

**Comment by balvinderz** [comment]
ok

**Comment by balvinderz** [comment]
https://github.com/flutter/plugins/pull/2974

**Comment by ditman** [comment]
Tagged and [published](https://pub.dev/packages/url_launcher_platform_interface/changelog#108) `url_launcher_platform_interface: ^1.0.8`

Next should probably be the `_web` version.

**Comment by ditman** [comment]
Both url_launcher and url_launcher_web have been published:
* https://pub.dev/packages/url_launcher/changelog#551
* https://pub.dev/packages/url_launcher_web/changelog#013

Thanks for the contribution @balvinderz!
